### PR TITLE
Added support for role based menus

### PR DESCRIPTION
**Pull request by bradymiller**
-Select menu in Administration->Users (Main Menu Role setting)
-Can select hard-coded standardized options (Standard, Answering Service, Front Office)
-Can also select custom options which are stored at:
  sites/\<site\>/documents/custom_menus/
  (currently is an example Custom.json there)
  (any json menu file stored there can be selected)

-The structures of the Answering Service and Front Office menus were
 written by Ray.

This is working and testing well.

### CODE REVIEW

File: interface/main/tabs/menu/menu_json.php [hunk 1]
```diff
@@ -1,41 +1,54 @@
...
+    // Note this is currently the standard mechanism
+
+    // Collect the selected menu by user
+    $mainMenuRole = "standard";
+    if (!empty($_SESSION['authUser'])) {
+        $sqlResult = sqlQuery("SELECT `main_menu_role` FROM `users` WHERE `username` = ?", array($_SESSION['authUser']));
+        if (!empty($sqlResult['main_menu_role'])) {
+            $mainMenuRole = $sqlResult['main_menu_role'];
+        }
+    }
```
Comment (bradymiller) [inline interface/main/tabs/menu/menu_json.php]: @MatthewVita ,
I am planning to migrate above code into services/UserService.php . I just wanted to make sure that would be in scope of the user service.

File: interface/usergroup/user_admin.php [hunk 2]
```diff
@@ -466,6 +343,36 @@ function authorized_clicked() {
...
+              echo "<option value='" . attr($menuCustom) . "' " . $selectedTag . ">";
+              // Drop the .json and translate the name
+              echo xlt(substr($menuCustom, 0, -5));
+              echo "</option>";
+            }
+          }
+          closedir($dHandle);
+        }
+      ?>
+    </select>
```
Comment (bradymiller) [inline interface/usergroup/user_admin.php]: @MatthewVita ,
I am planning to migrate above code (the creation of this select control) into services/UserService.php . I just wanted to make sure that would be in scope of the user service.

Comment (bradymiller) [inline interface/usergroup/user_admin.php]: Actually, maybe better to place this code block inside it's own mainMenuRole class.

File: interface/main/tabs/menu/menu_json.php [hunk 3]
```diff
@@ -1,41 +1,53 @@
...
 {
-    $menu_parsed = json_decode(file_get_contents(dirname(__FILE__) . "/menu_data.json"));
-    if (!$menu_parsed) die("\nJSON ERROR: " . json_last_error()); // debugging
+    // Use a json file to build menu rather than database
+    // Note this is currently the standard mechanism
+
+    // Collect the selected menu by user
+    $userService = new \services\UserService();
+    $user = $userService->getCurrentlyLoggedInUser();
+    $mainMenuRole = $user->getMainMenuRole();
```
Comment (bradymiller) [inline interface/main/tabs/menu/menu_json.php]: @MatthewVita ,
Hopefully I utilized the doctrine/service right. It is testing well.

**Comment by bradymiller** [comment]
I introduced a new class, so get to do the happy class dance:
happy-class-dance

**Comment by bradymiller** [comment]
Code is testing well and plugged it into doctrine/service/class. Please review.

File: library/menu/MainMenuRole.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ * @package OpenEMR
+ * @link    http://www.open-emr.org
+ * @author  Brady Miller <brady.g.miller@gmail.com>
+ * @copyright Copyright (c) 2017 Brady Miller <brady.g.miller@gmail.com>
+ * @license https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Menu;
+
```
Comment (bradymiller) [inline library/menu/MainMenuRole.php]: Need to add following here:
use services\UserService;

File: library/menu/MainMenuRole.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+{
+
+    /**
+     * Collect the MainMenuRole for logged in user.
+     *
+     * @return string Identifier for the MainMenuRole
+     */
+    static public function getMainMenuRole()
+    {
+        $userService = new \services\UserService();
```
Comment (bradymiller) [inline library/menu/MainMenuRole.php]: change to:
``````
$userService = new \UserService();

**Comment by bradymiller** [comment]
brief documentation placed here:
http://www.open-emr.org/wiki/index.php/Role_Based_Menus

**Comment by bradymiller** [comment]
Testing well, so will bring into codebase.
@mdsupport (is this mechanism something that could be leverages on your app code; ie. making menu specific for your app)

**Comment by sunsetsystems** [comment]
Looks like a cool feature. Any thoughts re ability for non-programmers to customize menus?

**Comment by bradymiller** [comment]
Hi @sunsetsystems ,

The hope is to have a drag drop mechanism to build/edit menus (which would be stored in the documents/custom_menus/ dir as the Custom.json menu is stored there). The nice thing is that the Standard menu will always have all the building blocks, so this is where the "menu" of menu building blocks can come from when building/editing a menu. All we need is a gui wiz to take it on :)

-brady

**Comment by robertdown** [comment]
@bradymiller I'll head the batcall....

**Pull request merged**